src/android/com/outsystems/firebase/analytics/validator/OSFANLEventParameterValidator.kt

```diff
@@ -0,0 +1,105 @@
+ * @return the String or null if the key isn't present
+ */
+class OSFANLEventParameterValidator private constructor(
+    private val requiredKeys: List<String>,
+    private val requireValueCurrency: Boolean,
+    private val numberKeys: List<String>,
+    private var maxLimit: Int? = null
+) {
+
+    class Builder {
+        private val requiredKeys = mutableListOf<String>()
+        private val numberKeys = mutableListOf<String>()
+        private var requireValueCurrency = false
+        private var maxLimit: Int? = null
+
+        fun required(vararg keys: OSFANLInputDataFieldKey) =
+            apply { keys.forEach { requiredKeys.add(it.json) } }
+        fun number(vararg keys: OSFANLInputDataFieldKey) =
+            apply { keys.forEach { numberKeys.add(it.json) } }
+        fun requireCurrencyValue() = apply { requireValueCurrency = true }
+        fun max(limit: Int) = apply { this.maxLimit = limit }
+
+        fun build() = OSFANLEventParameterValidator(
+            requiredKeys,
+            requireValueCurrency,
+            numberKeys,
+            maxLimit
+        )
+    }
+
+    /**
+     * Validates a JSONArray of parameters
+     * @param input the array of parameters to validate
+     * @return a bundle of parameters to send with the log event
+     */
+    fun validate(input: JSONArray): Bundle {
+
+        // validate maximum limit
+        maxLimit?.let {
+            if (input.length() >= it) throw OSFANLError.tooMany(EVENT_PARAMETERS.json, it)
+        }
+
+        val result = Bundle()
+        val parameterKeySet = mutableSetOf<String>()
+        var hasValue = false
+        var hasCurrency = false
+        for (i in 0 until input.length()) {
+            val parameter = input.getJSONObject(i)
+            val key = parameter.getString(KEY.json)
+            val value = parameter.getString(VALUE.json)
+
+            // validate type, if needed
+            if (numberKeys.contains(key) && value.toFloatOrNull() == null)
+                throw OSFANLError.invalidType(key, TYPE_NUMBER.json)
+
+            // validate duplicate keys
+            if (parameterKeySet.contains(key))
+                throw OSFANLError.duplicateKeys()
+
+            // validate value type
+            if (requireValueCurrency && key == VALUE.json && value.toFloatOrNull() == null)
+                throw OSFANLError.invalidType(VALUE.json, TYPE_NUMBER.json)
+
+            // search for value / currency
+            hasValue = hasValue || key == VALUE.json
+            hasCurrency = hasCurrency || key == CURRENCY.json
+
+            parameterKeySet.add(key)
+            result.putAny(key, value)
+        }
+
+        // validate required keys
+        requiredKeys.forEach {
+            if (!parameterKeySet.contains(it))
+                throw OSFANLError.missing(it)
+        }
+
+        // validate value / currency
+        // if value is present, currency is required
+        if (requireValueCurrency && hasValue && !hasCurrency)
```

Comment:
Should we perform this after the `requiredKeys.forEach` statement? This one appears to be more expensive to perform and could be avoided if the other, simpler condition is false.